File: api/proxy.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';

const ALLOWED_HOSTS = [
  'mp.weixin.qq.com',
  'mmbiz.qpic.cn',
  'mmbiz.url.cn',
  'mmsns.qpic.cn',
  'res.wx.qq.com',
  'szthumbnail.weixin.qq.com',
  'wqbrand.thumb.weixin.qq.com',
  'file.weixin.qq.com',
  'qqfile.weixin.qq.com',
];

function isAllowedUrl(urlStr: string): boolean {
  try {
    const u = new URL(urlStr);
    return ALLOWED_HOSTS.some(h => u.hostname === h || u.hostname.endsWith('.' + h));
  } catch {
    return false;
  }
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return res.status(204).send('');
  }

  if (req.method !== 'GET') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  const { url } = req.query;
  if (!url || typeof url !== 'string') {
    return res.status(400).json({ error: 'Missing url parameter' });
  }

  if (!isAllowedUrl(url)) {
    return res.status(400).json({ error: 'URL not allowed — only WeChat CDN domains are permitted' });
  }

  try {
    const resp = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': '*/*',
        'Accept-Language': 'zh-CN,zh;q=0.9',
        'Referer': 'https://mp.weixin.qq.com/',
      },
      redirect: 'follow',
    });

    if (!resp.ok) {
      return res.status(502).json({ error: `HTTP ${resp.status}` });
    }

    const contentType = resp.headers.get('content-type') || '';
    const buffer = await resp.arrayBuffer();

    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    res.setHeader('Cache-Control', 'public, max-age=86400');

    if (contentType.includes('text/html') || contentType.includes('application/json')) {
      let text = Buffer.from(buffer).toString('utf-8');
      if (!text.includes('charset') && !text.includes('encoding')) {
        try {
          const decoded = new TextDecoder('gbk').decode(Buffer.from(buffer));
          if (decoded.includes('微信公众号') || decoded.includes('wechat')) {
            text = decoded;
          }
        } catch (_) {}
      }
      return res.status(200).json({ contents: text });
    }

    res.setHeader('Content-Type', contentType);
    return res.status(200).send(Buffer.from(buffer));
  } catch (err: any) {
    return res.status(502).json({ error: err.message });
  }
}

File: proxy.py
#!/usr/bin/env python3
"""
Local WeChat article proxy server.
Run: python proxy.py
Then open index.html and use the proxy at http://localhost:51888/proxy?url=...
"""

import http.server
import urllib.request
import urllib.parse
import json
import sys
import gzip

PORT = 51888


class ProxyHandler(http.server.BaseHTTPRequestHandler):
    def log_message(self, fmt, *args):
        sys.stdout.write(f"[proxy] {args[0]}\n")
        sys.stdout.flush()

    def do_GET(self):
        if '/proxy' not in self.path:
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.end_headers()
            self.wfile.write(
                b"<html><body><h2>WeChat Proxy Server</h2>"
                b"<p>Use: <code>http://localhost:%d/proxy?url=ENCODED_URL</code></p>"
                b"<p>Example: <a href='/proxy?url=https%%3A%%2F%%2Fmp.weixin.qq.com%%2Fs%%2Fexample'>Test</a></p>"
                b"</body></html>" % PORT
            )
            return

        parsed = urllib.parse.urlparse(self.path)
        params = urllib.parse.parse_qs(parsed.query)

        if 'url' not in params:
            self.send_json(400, {"error": "missing url param"})
            return

        target = params['url'][0]
        if not target.startswith(('http://', 'https://')):
            self.send_json(400, {"error": "invalid url"})
            return

        try:
            req = urllib.request.Request(
                target,
                headers={
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Encoding": "gzip, deflate",
                    "Accept-Language": "zh-CN,zh;q=0.9",
                    "Referer": "https://mp.weixin.qq.com/",
                }
            )
            with urllib.request.urlopen(req, timeout=15) as resp:
                raw = resp.read()

                # Decompress gzip if needed
                try:
                    decoded = gzip.decompress(raw)
                except Exception:
                    decoded = raw

                # WeChat often lies about charset=utf-8 but sends GBK
                # Try UTF-8 strict; if it fails, fall back to GBK
                try:
                    content = decoded.decode('utf-8', errors='strict')
                except UnicodeDecodeError:
                    try:
                        content = decoded.decode('gbk', errors='replace')
                    except Exception:
                        content = decoded.decode('utf-8', errors='replace')

                self.send_json(200, {"contents": content})

        except Exception as e:
            self.send_json(502, {"error": str(e)})

    def send_json(self, code, data):
        body = json.dumps(data).encode('utf-8')
        self.send_response(code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", len(body))
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(body)


if __name__ == "__main__":
    print(f"Starting WeChat proxy server on http://localhost:{PORT}")
    print(f"Set this URL in index.html: http://localhost:{PORT}/proxy?url=")
    print("Press Ctrl+C to stop\n")
    server = http.server.HTTPServer(("0.0.0.0", PORT), ProxyHandler)
    server.serve_forever()
